### DISCUSSION TITLE
Reduce boilerplate and allow auto-generated properties with additional logic and initializers

### DISCUSSION THREAD

**Discussion by KillerBoogie** [dotnet/csharplang#5926] [Language Ideas] [answered]
Maybe my proposal comes from a lack of knowledge, but seeing it pragmatically, it is a natural consequence to me.

To be able to control access to fields properties are used as accessors. C# requires to define a private backing field and an extra access property, creating boilerplate code and making it less readable. Auto properties reduce that boilerplate code, but auto-implemented properties can't have any logic and only auto-implemented properties can have initializers. I don't understand why this limitation exists in the language. 

For example a `List `field where internals should be protected returning a clone requires a backing field. 

```
private List<string>_names = new ();
public List<string> Names => _names.DeepClone();
```

The compiler should hide the implementation of the accessors and create the backing field on its own.  A proposed short syntax is:

`public List<string> Names => Names.DeepClone() = new();`

or more specific

`public List<string> Names {get => Name.DeepClone(); private set { } } = new();`

This avoids the boilerplate code of the backing field, which is not needed. Every field that needs to be initialized extends the code unneccesarily. I don't want and should not to deal with that backing field in code. It would make field definitions more concise and clear. 

Language standards also declare a different naming convention for private fields. Therfore access to private backing fields and fields with auto-implemented properties look different. For example a constructor code look like

```
_names = names;
Duration = duration;
_descriptions=_descriptions;
MaxHeight = maxHeight;
_technicalInfo = technicalInfo;
```

The different naming conventions don't provide any information to the intention of the code. They are just a consequence of the language limitation.

There should be no properties, just fields with access control. Defining a field like

`public List<string>` could be equivalent to `public List<string> {get; set;}`.

**Comment by HaloFour** [comment, marked as answer]
I believe this should be covered by #140:

```cs
public List<string> Names { get => field.DeepClone(); } = new();
```

That would effectively compile down to:

```cs
private readonly List<string> _names = new List<string>();
public List<string> Names
{
    get
    {
        return _names.DeepClone();
    }
}
```

Here's a recent design meeting on the subject:

https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-02.md#initializers

**Reply by KillerBoogie** [reply]
I have not been able to read all of the comments of #140, yet. The field keyword is an acceptable solution to the problem and would make the code much better. 
I still ty to understand why the property name can't be used and the compiler would not being able to exchange `Names.DeepClone() `to `field.DeepClone()` automatically. Is it to differentiate if code inside the class accesses the backing field directly or via the declared setter and getter?

**Reply by CyrusNajmabadi** [reply]
>I still ty to understand why the property name can't be used and the compiler 

i'm not sure i understand your question.  A property name *can* be used already within a property.  It's entirely legal and well defined.  You may not like the behavior (it will recursively access the property), but it's certainly something that the language supports today (which means it is unlikely to change).

**Reply by KillerBoogie** [reply]
Thanks for the explanation. I was not aware of the recursive behavior. I have expected that the getter and setter are only executed when triggered from outside the class. I haven't run into this, because I had no need for it. With that existing behaviour the field keyword is a valuable and flexible solution.

**Comment by jnm2** [comment]
In your particular example, how do items get added to that list if there is no access to the backing field except from the property getter?

**Reply by KillerBoogie** [reply]
Items get addes with a method, like 
```
public Result AddItem(string item){
   //validation
   ...
   Names.Add(item);
   ...
}
```

In general I use private setters for proper encapsulation and then implement business cases with methods that alter the data and guarantee always valid objects. If the fields are just simple variable containers than a basic get; set; is already available.

My proposal is that the compiler can and should convert all access to a property in the class code to a call of the backing field automatically. Currently I need to manually create the backing field. I want the code less verbose and easier to understand. The differentiation between field and property never made sense to me. Renaming fields would also benefit from this syntax.

An option for parameters could be 

`public List<string> Names { get => Name.DeepClone(); set(string value) =>  Names.Add(value); }`

In any real business cases invariants and value checks need to be addressed. Therefore, I prefer methods for more control. Methods also allow me to return a Result object with proper error codes. For expectable errors (wrong user input) I want to return a Result and not throw exceptions, which are for unexpected errors (e.g. uncaught programming mistakes). Results are easy to combine, so that I can return multiple errors.

**Reply by jnm2** [reply]
But when you call `Names.Add(item);`, you're adding the item to a deep-cloned list. In your example, it's equivalent to doing `_names.DeepClone().Add(item)`. How does that cause `_names` to get an item added to it?

**Reply by KillerBoogie** [reply]
As I wrote above, I was not aware of the recursive behavior. I expected that `get `and `set ` are triggered only when called from outside the class.

**Reply by theunrepentantgeek** [reply]
> My proposal is that the compiler can and should convert all access to a property in the class code to a call of the backing field automatically

A common use of a setter is to enforce invariants, for example to ensure a string value is never null:

``` csharp
public string Name 
{
    get => _name;
    set => _name = value ?? "";
}
```

Having code within the class automatically bypass the setter logic would mean the invariant wouldn't be enforced within the class unless the logic was manually duplicated everywhere the property was set. This would be an error prone maintenance nightmare.

At the moment, developers have the choice to directly set the field `_name` (and bypass the invariant) or to set the property `Name` and have it enforced. They wouldn't have this choice if code internal to the type was silently redirected by the compiler.

**Reply by KillerBoogie** [reply]
Thanks for the explanation!

**Reply by jnm2** [reply]
@KillerBoogie I don't think there is recursion involved in what I'm asking about.

**Reply by KillerBoogie** [reply]
Recursive was the wrong word. I meant that the setter and getter code will be called for calls within the class when accessing the property.

**Reply by jnm2** [reply]
So, just to clarify: Even with #140, you'll still need to declare a private field in order to be able to add items to the list (later than the property initializer runs) in the given example.

**Comment by CyrusNajmabadi** [comment]
> The compiler should hide the implementation of the accessors and create the backing field on its own. A proposed short syntax is:

for your example:

```c#
public List<string> Names => Name.DeepClone() = new();
```

How does the compiler know what to do here?  What if `Name` has meaning already?  I'm not even sure what the rules are here. Is it that every identifier that is unbound becomes a field?  If so, what is the type of that field?

**Reply by jnm2** [reply]
Based on context, I think this was a typo and `Names.DeepClone()` was intended.

**Reply by KillerBoogie** [reply]
Yes, it was a typo. It should have been `Names.DeepClone()`, the same identifier name as the property.
